src/System.Net.Security/src/System/Net/Security/SniHelper.cs

```diff
@@ -0,0 +1,377 @@
+            //         ...
+            //     } body;
+            // } Handshake;
+            const int HandshakeTypeOffset = 0;
+            const int ClientHelloLengthOffset = HandshakeTypeOffset + sizeof(HandshakeType);
+            const int ClientHelloOffset = ClientHelloLengthOffset + UInt24Size;
+
+            if (sslHandshake.Length < ClientHelloOffset || (HandshakeType)sslHandshake[HandshakeTypeOffset] != HandshakeType.ClientHello)
+            {
+                return null;
+            }
+
+            int clientHelloLength = ReadUInt24BigEndian(sslHandshake.Slice(ClientHelloLengthOffset));
+            ReadOnlySpan<byte> clientHello = sslHandshake.Slice(ClientHelloOffset);
+
+            if (clientHello.Length != clientHelloLength)
+            {
+                return null;
+            }
+
+            return GetSniFromClientHello(clientHello);
+        }
+
+        private static string GetSniFromClientHello(ReadOnlySpan<byte> clientHello)
+        {
+            // Basic structure: https://tools.ietf.org/html/rfc6101#section-5.6.1.2
+            // Extended structure: https://tools.ietf.org/html/rfc3546#section-2.1
+            // struct {
+            //     ProtocolVersion client_version; // 2x uint8
+            //     Random random; // 32 bytes
+            //     SessionID session_id; // opaque type
+            //     CipherSuite cipher_suites<2..2^16-1>; // opaque type
+            //     CompressionMethod compression_methods<1..2^8-1>; // opaque type
+            //     Extension client_hello_extension_list<0..2^16-1>;
+            // } ClientHello;
+            ReadOnlySpan<byte> p = SkipBytes(clientHello, ProtocolVersionSize + RandomSize);
+
+            // Skip SessionID (max size 32 => size fits in 1 byte)
+            p = SkipOpaqueType1(p);
+
+            // Skip cipher suites (max size 2^16-1 => size fits in 2 bytes)
+            p = SkipOpaqueType2(p, out _);
+
+            // Skip compression methods (max size 2^8-1 => size fits in 1 byte)
+            p = SkipOpaqueType1(p);
+
+            // is invalid structure or no extensions?
+            if (p.IsEmpty)
+            {
+                return null;
+            }
+
+            // client_hello_extension_list (max size 2^16-1 => size fits in 2 bytes)
+            int extensionListLength = BinaryPrimitives.ReadUInt16BigEndian(p);
+            p = SkipBytes(p, sizeof(ushort));
+
+            if (extensionListLength != p.Length)
+            {
+                return null;
+            }
+
+            string ret = null;
+            while (!p.IsEmpty)
+            {
+                bool invalid;
+                string sni = GetSniFromExtension(p, out p, out invalid);
+                if (invalid)
+                {
+                    return null;
+                }
+
+                if (ret != null && sni != null)
+                {
+                    Debug.Assert(false, "More than 1 name found.");
```

Comment:
Is it expected that a web server would ever possibly run a checked build of corefx in their production environment? If so, `Debug.Assert` allows a malicious payload to crash the web service. If this isn't a legitimate scenario, disregard this comment.